[rocprofiler-systems]Documentation edits to MPI rank profiling added (#7396)
ROCm/rocm-systems: projects/rocprofiler-systems/docs/how-to/communication-runtime-profiling.rst

* When full MPI support is enabled (``ROCPROFSYS_USE_MPI=ON``), output files are labeled with the ``MPI_COMM_WORLD`` rank ID
* The ``ROCPROFSYS_USE_PID`` setting controls whether process IDs or MPI rank IDs are used for output labeling

For detailed information on building ROCm Systems Profiler with MPI support, see the :doc:`installation guide <../install/install>`.

Selective rank profiling
-------------------------

When running large-scale MPI jobs, you might not always want to collect profiling data and console output from all MPI ranks.
ROCm Systems Profiler provides two independent rank-filtering options:

- ``--rank-filter-output``: Only the specified MPI ranks produce profile and trace output files. Other ranks run normally but don't generate output files. The corresponding configuration setting is ``ROCPROFSYS_RANK_FILTER_OUTPUT``.

> - ``--rank-filter-logs``: Only the listed MPI ranks emit console output. Other ranks run normally but don't generate output files. The corresponding configuration setting is ``ROCPROFSYS_RANK_FILTER_LOGS``.

Reviewer: ```suggestion
- ``--rank-filter-logs``: Only the listed MPI ranks emit console output. Other ranks run normally but don't generate console output. The corresponding configuration setting is ``ROCPROFSYS_RANK_FILTER_LOGS``.
```

    # so only rank 1 produces output
    mpirun -n 16 rocprof-sys-sample --rank-filter-output "1,100" -- <application_path>

> ROCm Systems Profiler automatically detects each process's MPI rank by checking these _`rank identification` environment variables:

Reviewer 2: I think we should mention that the following environment variables are checked **in priority order**, and the
first one that is set to a valid non-negative integer is used as the rank. @lloginov-amd ?

Reviewer: I agree, let's add immediately after the list of variables something like:

"These environment variables are checked in the listed order top-to-bottom, and the first one that is set to a non-negative integer is used as the rank."

- ``MPI_RANKID``
- ``PMI_RANK`` 
- ``MV2_COMM_WORLD_RANK``
- ``OMPI_COMM_WORLD_RANK``

> - ``SLURM_PROCID``

Reviewer 2: The custom filter set with `--rank-filter-id` or `ROCPROFSYS_RANK_FILTER_ID` takes precedence over all of the variables above.

Reviewer: Maybe we can omit this - we already have it mentioned later in "Custom MPI environment variables" section. If a user need custom env vars - they will visit this section and see this note.

Reviewer 2: If so, then we need to explicitly mention this in the below section for custom mpi environment that rank_filter_id has highest precedence.

Reviewer: We currently have this line in "Custom MPI environment variables" below:

"The ``--rank-filter-id`` will take precedence over automatic detection for both filters."

Or did you mean something else?

The profiler uses the detected rank number to determine whether that process should produce output files or console logs based on your filter settings. If rank detection fails using these standard variables, both filters are disabled and all ranks produce the output.

Similarly, if the detected rank itself is out of the range ``[0, world_size - 1]``, the filtering is disabled for that rank and all ranks produce the output.

.. note::

>    For complete console output suppression, use the ``ROCPROFSYS_RANK_FILTER_LOGS`` environment variable instead of the ``--rank-filter-logs`` CLI option. The command-line option may still show some log messages during startup. Also note that errors and critical messages are emitted for all ranks regardless of rank-based filtering option.

Reviewer 2: I don't have a full understanding about how we differentiate the names `Configuration setting` and `Environment variable` in our current documentation, however in true sense all `ROCPROFSYS_*` are environment variables too, but we refer to `ROCPROFSYS_*` more as a configuration setting.

Author: Seems like they are being used interchangeably. We should use a uniform convention across the document to minimize ambiguity.

Reviewer: 1. word "complete" here would be untrue - there are some logs being shown even with `ROCPROFSYS_RANK_FILTER_LOGS`, though very little. I suggest reverting to "maximum", but maybe there is a better word for that. It's just that "complete" is not exact here.

2. I suggest we use `configuration setting` here. The rest of this section uses `configuration setting` for `ROCPROFSYS_*`, so let's also use it here for consistency. This way, `Environment variable` is reserved for variables that are NOT profiler-specific settings (like variables used for rank detection).

```suggestion
   For maximum console output suppression, use the ``ROCPROFSYS_RANK_FILTER_LOGS`` configuration setting instead of the ``--rank-filter-logs`` CLI option. The command-line option may show more log messages during startup. Also note that errors and critical messages are emitted for all ranks regardless of rank-based filtering option.
```

   .. code-block:: bash

       # Show console output only from rank 0
       export ROCPROFSYS_RANK_FILTER_LOGS=0
       mpirun -n 16 rocprof-sys-sample -- <application_path>

> For more details on selective rank profiling using ``rocprofv3``, refer to :doc:`Using rocprofv3 with MPI <rocprofiler-sdk:how-to/using-rocprofv3-with-mpi>`.

Reviewer 2: I am not certain if the mention about `rocprofv3` tool is necessary in this context.

Author: This reference was suggested by @dgaliffiAMD . Can remove it if not relevant.

Reviewer 3: I think my suggestion was misinterpreted. I apologize for the misunderstanding, but let's drop it. 
```suggestion
```

Custom MPI environment variables
----------------------------------

> For mixed environments or non-standard MPI configurations, you can specify custom environment variables for rank detection.

Reviewer: Let's add the env var setting info to this first sentence:

"For mixed environments or non-standard MPI configurations, you can specify custom environment variables for rank detection using `--rank-filter-id` CLI option or corresponding configuration setting `ROCPROFSYS_RANK_FILTER_LOGS`".

> When using custom environment variables, ``--rank-filter-id`` must be specified together with at least one of the filter:  ``--rank-filter-output`` or ``--rank-filter-logs``.

Reviewer 2: Somewhere in this section we can also mention about the config setting `ROCPROFSYS_RANK_FILTER_ID` which is the same as `--rank-filter-id`. @lloginov-amd

Reviewer: Absolutely! I suggest adding it to the sentence above - see my comment there.

Author: Which is the correct configuration setting ``ROCPROFSYS_RANK_FILTER_LOGS`` or ``ROCPROFSYS_RANK_FILTER_ID``?

Reviewer 2: All are acceptable. Each setting serves different purposes.

Reviewer 2: Above comment is about `ROCPROFSYS_RANK_FILTER_ID` meaning the same setting as option `--rank-filter-id`. Similar to `ROCPROFSYS_RANK_FILTER_LOGS` same as option `--rank-filter-output`.

Reviewer: @sputhala-amd, I think there is a typo: `ROCPROFSYS_RANK_FILTER_LOGS `corresponds to `--rank-filter-logs`, not `--rank-filter-output`. So:

`--rank-filter-id` = `ROCPROFSYS_RANK_FILTER_ID`
`--rank-filter-output` = `ROCPROFSYS_RANK_FILTER_OUTPUT`
`--rank-filter-logs` = `ROCPROFSYS_RANK_FILTER_LOGS`

Custom MPI environment variables
----------------------------------

> For mixed environments or non-standard MPI configurations, you can specify custom environment variables for rank detection using ``--rank-filter-id`` CLI option or corresponding configuration setting ``ROCPROFSYS_RANK_FILTER_LOGS``.

Reviewer: Should be ROCPROFSYS_RANK_FILTER_ID here

```suggestion
For mixed environments or non-standard MPI configurations, you can specify custom environment variables for rank detection using ``--rank-filter-id`` CLI option or corresponding configuration setting ``ROCPROFSYS_RANK_FILTER_ID``.
```
